### PR TITLE
add ecommerce blog post part 3

### PR DESCRIPTION
**Pull request by vkarpov15**
## What does this PR do?

Added part 3 of the ecommerce blog series from here: https://github.com/temporalio/temporal-ecommerce/pull/10

### CODE REVIEW

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+}
+```
+
+## Moving On
+
+Temporal's testing environment makes unit testing Workflows easy.
+Sending signals and queries to the currently running Workflow is straightforward with `RegisterDelayedCallback()`.
+Most importantly, Temporal's testing environment provides utilities for mocking Activities and testing logic that executes after a delay.
+That makes it to unit test Workflows that depend on external services or Workflows that involve long timeouts.
+In the next blog post in this series, you'll learn how to built a RESTful API on top of a Temporal Workflow.
```
Comment (sw-yx) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:366]:
```suggestion
In the next blog post in this series, you'll learn how to build a RESTful API on top of a Temporal Workflow.
```

**Comment by sw-yx** [comment]
this looks fine to me, not really sure how to make it more engaging but it is at least very educational. @flossypurse ?

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+In [Part 1](/blog/build-an-ecommerce-app-with-temporal-part-1) and [Part 2](/blog/build-an-ecommerce-app-with-temporal-part-2-reminder-emails), you built out a shopping cart with an abandoned cart email notification using long-lived Workflows.
+Using Workflows and Activities, you can easily build features that would be tricky in a traditional RESTful API, like sending an email reminder when a user hasn't touched their cart in a while.
+
+Temporal Workflows do more than make working with time easy, they also make your code easier to test.
+Temporal provides testing utilities that help you stub out external services and programmatically advance time, which lets you [unit test your Workflows](/blog/descript-case-study/#to-code-or-not-to-code).
+In this blog post, I'll demonstrate how to use Temporal's testing utilities to write fast unit tests for the shopping cart from Part 1 and Part 2.
+
+## Testing Setup
+
+Below is a basic setup for testing a Temporal workflow using `go test` and [Testify](https://github.com/stretchr/testify) based on [Temporal's Go testing docs](/docs/go/testing).
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:31]:
```suggestion
Below is a basic setup for testing a Temporal Workflow using `go test` and [Testify](https://github.com/stretchr/testify) based on [Temporal's Go testing docs](/docs/go/testing).
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+	s.env.AssertExpectations(s.T())
+}
+
+func TestUnitTestSuite(t *testing.T) {
+	suite.Run(t, new(UnitTestSuite))
+}
+```
+
+The most important property is the `env` property, which is an instance of [Temporal's `TestWorkflowEnvironment` struct](https://pkg.go.dev/go.temporal.io/temporal/internal#TestWorkflowEnvironment).
+A `TestWorkflowEnvironment` provides utilities for testing Workflows, including executing Workflows, mocking Activities, and signaling and querying test Workflows.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:75]:
```suggestion
A `TestWorkflowEnvironment` provides utilities for testing Workflows, including executing Workflows, mocking Activities, and Signaling and Querying test Workflows.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+func (s *IntegrationTestSuite) AfterTest(suiteName, testName string) {
+	s.env.AssertExpectations(s.T())
+}
+```
+
+## Querying Workflows in tests
+
+Remember that, in this app, a shopping cart is a Workflow.
+To get the current state of the shopping cart, you send a query to the Workflow, and to update the cart you send a signal to the Workflow.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:115]:
```suggestion
To get the current state of the shopping cart, you send a Query to the Workflow, and to update the cart you send a Signal to the Workflow.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+
+The below code shows how you can use `env.QueryWorkflow()` to send a query to the shopping cart Workflow.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:117]:
```suggestion
The below code shows how you can use `env.QueryWorkflow()` to send a Query to the shopping cart Workflow.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+  // Send a query to the Workflow and assert that the shopping cart is still empty
+	res, err := s.env.QueryWorkflow("getCart")
+	s.NoError(err)
+	err = res.Get(&cart)
+	s.NoError(err)
+	s.Equal(0, len(cart.Items))
+}
+```
+
+Note that the above code queries the Workflow _after the Workflow is done_.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:138]:
```suggestion
Note that the above code Queries the Workflow _after the Workflow is done_.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+In order to interact with the Workflow via queries and signals while the Workflow is running, you should use the test environment's `RegisterDelayedCallback()` function as shown below.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:139]:
```suggestion
In order to interact with the Workflow via Queries and Signals while the Workflow is running, you should use the test environment's `RegisterDelayedCallback()` function as shown below.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+		s.Equal(len(cart.Items), 0)
+	}, time.Millisecond*1)
+
+	s.env.ExecuteWorkflow(CartWorkflow, cart)
+
+	s.True(s.env.IsWorkflowCompleted())
+}
+```
+
+You can query a Workflow after it is completed, but you can't signal a Workflow after it is completed.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:161]:
```suggestion
You can Query a Workflow after it is completed, but you can't Signal a Workflow after it is completed.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+## Signaling Workflows in tests
+
+So in order to signal a Workflow from your tests, you need to use `RegisterDelayedCallback()`.
+Just remember that signaling is asynchronous, so you need to add a separate `RegisterDelayedCallback()` to read the result of your signal using a query.
+For example, below is a test case for the `AddToCart()` method.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:167]:
```suggestion
So in order to Signal a Workflow from your tests, you need to use `RegisterDelayedCallback()`.
Just remember that Signaling is asynchronous, so you need to add a separate `RegisterDelayedCallback()` to read the result of your Signal using a Query.
For example, below is a test case for the `AddToCart()` method.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    s.Equal(1, len(cart.Items))
+    s.Equal(1, cart.Items[0].Quantity)
+	}, time.Millisecond*2)
+
+	s.env.ExecuteWorkflow(CartWorkflow, cart)
+
+	s.True(s.env.IsWorkflowCompleted())
+}
+```
+### Sending multiple signals to Workflows in tests
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:206]:
```suggestion
### Sending multiple Signals to Workflows in tests
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+
+Similarly, if you want to send multiple signals, you should put them in separate `RegisterDelayedCallback()` calls.
+However, you can move any queries that don't have any signals after them to after the `ExecuteWorkflow()` call.
+For example, below is a test case for the `RemoveFromCart()` method.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:210]:
```suggestion
Similarly, if you want to send multiple Signals, you should put them in separate `RegisterDelayedCallback()` calls.
However, you can move any Queries that don't have any Signals after them to be after the `ExecuteWorkflow()` call.
For example, below is a test case for the `RemoveFromCart()` method.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+}
+```
+
+This covers testing the basic functionality of adding items to and removing items from the shopping cart.
+But what about testing more sophisticated features, like testing that the Workflow sends an abandoned cart email after 10 minutes?
+
+## Mocking Activities and Controlling Time
+
+Temporal's test environment makes it easy to mock activities, replacing them with a stubbed out function.
+For example, the below test asserts that sending a checkout signal calls the `CreateStripeCharge` activity with the correct receipt email using the test environment's `OnActivity()` function.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:260]:
```suggestion
Temporal's test environment makes it easy to mock Activities, replacing them with a stubbed out function.
For example, the below test asserts that sending a checkout Signal calls the `CreateStripeCharge` Activity with the correct receipt email using the test environment's `OnActivity()` function.
```

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+	s.env.ExecuteWorkflow(CartWorkflow, cart)
+
+	s.True(s.env.IsWorkflowCompleted())
+}
+```
+
+## Moving On
+
+Temporal's testing environment makes unit testing Workflows easy.
+Sending signals and queries to the currently running Workflow is straightforward with `RegisterDelayedCallback()`.
```
Comment (flossypurse) [inline blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md:363]:
```suggestion
Sending Signals and Queries to the currently running Workflow is straightforward with `RegisterDelayedCallback()`.
```

**Review by flossypurse — COMMENTED** [review]
Awesome.
There is a lot of great info here.

Regarding @sw-yx  comment about increasing engagement, I might suggest putting in a "Lore check" like we do in these tutorials: https://docs.temporal.io/docs/go/run-your-first-app-tutorial/#lore-check

File: blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+	}, time.Millisecond*2)
+
+	s.env.ExecuteWorkflow(CartWorkflow, cart)
+
+	s.True(s.env.IsWorkflowCompleted())
+}
+```
+### Sending multiple Signals to Workflows in tests
+
+Similarly, if you want to send multiple Signals, you should put them in separate `RegisterDelayedCallback()` calls.
```
[the remark this answers is not in the document]
Comment (vkarpov15) [reply blog/2021-06-23-build-an-ecommerce-app-with-temporal-part-3.md]: Good point. You can send multiple signals in one `RegisterDelayedCallback()`, you just can't read the updated state using a Query in the same `RegisterDelayedCallback()`. I updated this in 2231c53 :+1:

**Review by sw-yx — APPROVED** [review]
lgtm, remember to change dates

**Pull request merged**